runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -7,102 +7,135 @@ package aws.smithy.kotlin.runtime.http.interceptors
+    requestChecksumRequired: Boolean,
+    requestChecksumCalculation: HttpChecksumConfigOption?,
+    requestChecksumAlgorithm: String?,
 ) : AbstractChecksumInterceptor() {
-    private var checksumAlgorithmName: String? = null
 
-    @Deprecated("readAfterSerialization is no longer used")
+    // FIXME: Remove in next minor version bump
+    @Deprecated("Old constructor is no longer used but it's kept for backwards compatibility")
+    public constructor() : this(
+        false,
+        HttpChecksumConfigOption.WHEN_REQUIRED,
+        null,
+    )
+
+    // FIXME: Remove in next minor version bump
+    @Deprecated("Old constructor is no longer used but it's kept for backwards compatibility")
+    public constructor(
+        checksumAlgorithmNameInitializer: ((I) -> String?)? = null,
+    ) : this(
+        false,
+        HttpChecksumConfigOption.WHEN_REQUIRED,
+        null,
+    )
+
+    private val checksumHeader = buildString {
+        append("x-amz-checksum-")
+        append(requestChecksumAlgorithm?.lowercase() ?: "crc32")
+    }
+
+    private val checksumAlgorithm = requestChecksumAlgorithm
+        ?.toHashFunctionOrThrow()
+        ?.takeIf { it.isSupportedForFlexibleChecksums }
+        ?: (Crc32().takeIf { requestChecksumRequired || requestChecksumCalculation == HttpChecksumConfigOption.WHEN_SUPPORTED })
+
+    // TODO: Remove in next minor version bump
+    @Deprecated("readAfterSerialization is no longer used but can't be removed due to backwards incompatibility")
     override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) { }
 
     override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
         val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor<I>>()
 
-        @Suppress("UNCHECKED_CAST")
-        val input = context.request as I
-        checksumAlgorithmName = checksumAlgorithmNameInitializer?.invoke(input) ?: context.executionContext.getOrNull(HttpOperationContext.ChecksumAlgorithm)
+        context.protocolRequest.userProvidedChecksumHeader(logger)?.let {
+            logger.debug { "Checksum was supplied via header, skipping checksum calculation" }
 
-        checksumAlgorithmName ?: run {
-            logger.debug { "no checksum algorithm specified, skipping flexible checksums processing" }
-            return context.protocolRequest
+            val request = context.protocolRequest.toBuilder()
+            request.headers.removeAllChecksumHeadersExcept(it)
+            return request.build()
         }
 
-        val req = context.protocolRequest.toBuilder()
-
-        check(context.protocolRequest.body !is HttpBody.Empty) {
-            "Can't calculate the checksum of an empty body"
+        if (checksumAlgorithm == null) {
+            logger.debug { "A checksum algorithm isn't selected and checksum calculation isn't required, skipping checksum calculation" }
```

Comment:
We could also fall into this case when a user _does_ select a checksum algorithm but they also configure `WHEN_REQUIRED`. 

I think a better log would be:
"A checksum algorithm isn't selected **or** checksum calculation isn't required, skipping checksum calculation"